app/assets/src/components/visualizations/heatmap/Heatmap.js

```diff
@@ -400,6 +439,165 @@ export default class Heatmap {
           this.options.spacing}
       )`
     );
+
+    // Set up space bar+click to pan behavior.
+    d3
+      .select("body")
+      .on("keydown", () => {
+        if (d3.event.code === "Space") {
+          this.spacePressed = true;
+        }
+      })
+      .on("keyup", () => {
+        if (d3.event.code === "Space") {
+          this.spacePressed = false;
+        }
+      });
+
+    this.svg
+      .on("mousedown", () => {
+        this.mouseDown = true;
+        if (this.spacePressed) {
+          this.mouseX = d3.event.clientX;
+          this.mouseY = d3.event.clientY;
+          d3.event.preventDefault();
+        }
+      })
+      .on("mousemove", this.drag.bind(this))
+      .on("mouseup", () => {
+        this.mouseDown = false;
+      });
+
+    // Set up scrolling behavior.
+    this.gCells.on("wheel.zoom", this.scroll.bind(this));
+  }
+
+  drag() {
+    if (this.mouseDown && this.spacePressed) {
+      this.pan(d3.event.clientX - this.mouseX, d3.event.clientY - this.mouseY);
+      this.mouseX = d3.event.clientX;
+      this.mouseY = d3.event.clientY;
+    }
+  }
+
+  scroll() {
+    this.pan(d3.event.wheelDeltaX, d3.event.wheelDeltaY);
+    d3.event.stopPropagation();
+  }
+
+  pan(deltaX, deltaY) {
```

Comment:
This function has a lot of duplicated code with placeContainers. The only difference is that placeContainers uses `0` while this method uses `labelsDx` and `labelsDy` when placing various containers. We discussed in person factoring out the place code for the two sets of d3 elements that "move" when the heatmap is shifted, which will help de-dupe the code.